File: src/main/java/com/w11k/examples/kotlin/library/Library.java
package com.w11k.examples.kotlin.library;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

public class Library {
    private List<Item> catalog = new ArrayList<>();

    public Library() {
    }

    public Library(List<Item> catalog) {
        this.catalog = catalog;
    }

    public Library addItem(Item item) {
        this.catalog.add(item);
        return this;
    }

    public List<Item> getCatalog() {
        return this.catalog.stream().collect(Collectors.toList());
    }
}

File: src/main/kotlin/com/w11k/examples/kotlin/library/Library.kt
package com.w11k.examples.kotlin.library

abstract class Item(open val inventoryId: String) {
}

data class Book(
    override val inventoryId: String,
    val title: String,
    val year: String,
    val author: Person,
    val isHardcover: Boolean = true
) : Item(inventoryId)

data class Article(
    override val inventoryId: String,
    val title: String,
    val author: Person,
    val abstract: String? = null
): Item(inventoryId)

data class Person(val name: String, val gender: String)